packages/rum-react/src/get-with-transaction.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * MIT License
+ *
+ * Copyright (c) 2017-present, Elasticsearch BV
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+/*
+    Usage:
+        - Pure function: `withTransaction('name','route-change')(Component)`
+        - As a decorator: `@withTransaction('name','route-change')`
+ */
+
+import React from 'react'
+import hoistStatics from 'hoist-non-react-statics'
+
+function getWithTransaction(apm) {
+  if (!apm) {
+  }
+  return function withTransaction(name, type) {
+    if (typeof name !== 'string') {
+      //todo: warn user about the usage
+    }
+    return function(WrappedComponent) {
+      class ApmComponent extends React.Component {
+        constructor(props) {
+          super(props)
+          /*
+                        We need to start the transaction in constructor because otherwise,
+                        we won't be able to capture what happens in componentDidMount of child components.
+                    */
+          this.transaction = apm.startTransaction(name, type)
```

Comment:
Did you test the transaction timing when the component is rendered lazily?

Reply:
Can you elaborate on the scenario you have in mind?

Reply:
React decides when the component tree should be mounted in to the DOM, So I was wondering if we are capturing the transaction time properly during these scenarios - https://github.com/reactjs/rfcs/blob/gaearon-patch-2/text/0000-lazy.md

Reply:
Have you checked about this scenario?